[openimageio] Update to v2.3.17.0, revise dependencies and export

dg0yt opened the pull request:
Systematic review of dependencies lookup and export, based on log files and exported config.

- #### What does your PR fix?
  Update to 2.3.17.0. Resolves #24337.
  Fixes #13621.
  Fixes broken imports and exports.
  Cleanup.
  Guards against requesting the opencolorio feature unless using an old opencolorio port version which still uses the openexr Imath. (Not a regression in this update. Eventually to be fixed by updating openexr.)
  [libheif] Burns dll import into header. Fixes mingw builds.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  unchanged, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  yes

dg0yt committed:
Consolidate patches

dg0yt committed:
Revise OpenJPEG patching

dg0yt committed:
Revise PNG patching

dg0yt committed:
Revise libheif patching

dg0yt committed:
Revise libsquish patching

dg0yt committed:
Misc cleanup

dg0yt committed:
Update copyright, add usage

dg0yt committed:
Trim dependencies

dg0yt committed:
OpenImageIO features CI test

Neumann-A commented on ports/openimageio/portfile.cmake:
```diff
@@ -16,43 +15,63 @@ file(REMOVE "${SOURCE_PATH}/src/cmake/modules/FindLibRaw.cmake"
+        -DCMAKE_REQUIRE_FIND_PACKAGE_Libsquish=ON
+        -DCMAKE_REQUIRE_FIND_PACKAGE_PNG=ON
+        -DCMAKE_REQUIRE_FIND_PACKAGE_Robinmap=ON
+    MAYBE_UNUSED_VARIABLES
+        CMAKE_REQUIRE_FIND_PACKAGE_FFmpeg
+        CMAKE_REQUIRE_FIND_PACKAGE_Freetype
+        CMAKE_REQUIRE_FIND_PACKAGE_GIF
+        CMAKE_REQUIRE_FIND_PACKAGE_LibRaw
+        CMAKE_REQUIRE_FIND_PACKAGE_OpenColorIO
+        CMAKE_REQUIRE_FIND_PACKAGE_OpenCV
+        CMAKE_REQUIRE_FIND_PACKAGE_OpenJPEG
+        CMAKE_REQUIRE_FIND_PACKAGE_Python
+        CMAKE_REQUIRE_FIND_PACKAGE_WebP
```
maybe this should be fixed in `vcpkg_configure_cmake` first.  If those variables are off it should simply be ignored. (I ignore them in Qt automatically depending on value. Same for `CMAKE_DISABLE_`)

dg0yt committed:
Revise opencv dependency

dg0yt committed:
Remove find modules which might clash with opencv

dg0yt committed:
Use opencolorio config and target

dg0yt committed:
Fix exported Imath dependency

dg0yt committed:
Drop unused deps from patches and config

dg0yt committed:
Handle DLL import in installed header

dg0yt commented:
So, openimageio and opencolorio cannot be used together in their current configurations:
~~~
duplicate symbol 'printBits(char*, float)' in:
    /Users/vagrant/Data/installed/x64-osx/debug/lib/libHalf-2_5_d.a(half.cpp.o)
    /Users/vagrant/Data/installed/x64-osx/debug/lib/libImath-3_1_d.a(half.cpp.o)
duplicate symbol 'printBits(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, float)' in:
    /Users/vagrant/Data/installed/x64-osx/debug/lib/libHalf-2_5_d.a(half.cpp.o)
    /Users/vagrant/Data/installed/x64-osx/debug/lib/libImath-3_1_d.a(half.cpp.o)
~~~

This is one "half" from imath, and another one from openexr. IIRC opencolorio might use the one from openexr, too. But I don't know the roadmao of these projects.

dg0yt commented:
> This is one "half" from imath, and another one from openexr. IIRC opencolorio might use the one from openexr, too. But I don't know the roadmap of these projects.

Well, there is #20957.

dg0yt committed:
Fix non-required find_package(OpenColorIO)

dg0yt committed:
Omit unrelocatable script

dg0yt committed:
Installation cleanup

dg0yt commented:
osx: Python problem, `-l_PYTHON_INTERFACE_LIBS-NOTFOUND` 
linux: OpenCV problem, `A required library with LAPACK API not found.`
windows-static: libheif exports problem (attempted to fix it here, but not yet there).

Adela0814 commented:
Pinging @dg0yt  for response. Is work still being done for this PR?

dg0yt commented:
I will pick this up again.

dg0yt committed:
Update openimageio to 2.3.17.0

dg0yt commented:
So, the `python3` wrapper blindly assumes an `INTERFACE_LINK_LIBRARIES` property for `@PythonFinder_PREFIX@::Python`... 😠

dg0yt commented:
From expanded trace at https://dev.azure.com/vcpkg/public/_build/results?buildId=74899&view=results:
The x64-linux error starts at a `FindLAPACK.cmake` failure deep down in package dependencies:
~~~
: && /usr/bin/cc -fPIC -DCHECK_FUNCTION_EXISTS=cheev_  CMakeFiles/cmTC_9da28.dir/CheckFunctionExists.c.o -o cmTC_9da28  /mnt/vcpkg-ci/installed/x64-linux/debug/lib/liblapack.a  /usr/lib/x86_64-linux-gnu/libm.a  /usr/lib/x86_64-linux-gnu/libgfortran.so.5  /mnt/vcpkg-ci/installed/x64-linux/debug/lib/libopenblas.a && :
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libm-2.31.a(e_logf.o): in function `__logf_ifunc':
(.text+0x146): undefined reference to `_dl_x86_cpu_features'
~~~
It doesn't link with `--static` (`VCPKG_CRT_LINKAGE dynamic`), but uses a static `libm` found via `find_library`. AFAIU the combination of static `libm` with shared `libc` is not allowed. Cf. https://bugzilla.redhat.com/show_bug.cgi?id=1433347
CC @Neumann-A 

This failure causes the ceres package to not restore `CMAKE_MODULE_PATH` which in turn makes `find_package(Robinmap)` fail.

Neumann-A commented:
> Trying to pass set LAPACK_m_LIBRARY to m or -lm didn't help to pass check_function_exists.

Tried to change it to `m.so` instead ?

dg0yt commented:
> Tried to change it to `m.so` instead ?

`find_library(... libm.so)` provides the right lib.
However, `LAPACK_m_library` is just the tip of the iceberg. For openimageio, there is `libm.a´ from qt5 (many times) and from ffmpeg.

Neumann-A commented:
So it is more of a general problem than a lapack specific one.

dg0yt commented:
> So it is more of a general problem than a lapack specific one.

AFAIU it is a very general problem with the official non-windows triplets in vcpkg, due to the combination of static library linkage with dynamic CRT linkage.

dg0yt committed:
LINKSTATIC breaks CRT lib linkage

dg0yt commented:
To make the problem even more general, this is a list of libs in Ubuntu 18.04's package `libc6-dev` which have both shared and static variants.

~~~
/usr/lib/x86_64-linux-gnu/libanl.a
/usr/lib/x86_64-linux-gnu/libanl.so
/usr/lib/x86_64-linux-gnu/libBrokenLocale.a
/usr/lib/x86_64-linux-gnu/libBrokenLocale.so
/usr/lib/x86_64-linux-gnu/libc.a
/usr/lib/x86_64-linux-gnu/libc.so
/usr/lib/x86_64-linux-gnu/libcrypt.a
/usr/lib/x86_64-linux-gnu/libcrypt.so
/usr/lib/x86_64-linux-gnu/libdl.a
/usr/lib/x86_64-linux-gnu/libdl.so
/usr/lib/x86_64-linux-gnu/libm.a
/usr/lib/x86_64-linux-gnu/libm.so
/usr/lib/x86_64-linux-gnu/libmvec.a
/usr/lib/x86_64-linux-gnu/libmvec.so
/usr/lib/x86_64-linux-gnu/libnsl.a
/usr/lib/x86_64-linux-gnu/libnsl.so
/usr/lib/x86_64-linux-gnu/libpthread.a
/usr/lib/x86_64-linux-gnu/libpthread.so
/usr/lib/x86_64-linux-gnu/libresolv.a
/usr/lib/x86_64-linux-gnu/libresolv.so
/usr/lib/x86_64-linux-gnu/librt.a
/usr/lib/x86_64-linux-gnu/librt.so
/usr/lib/x86_64-linux-gnu/libutil.a
/usr/lib/x86_64-linux-gnu/libutil.so
~~~

IIUC they all belong to the CRT, requiring `VCPKG_CRT_LINKAGE`, thus cannot be left to an uninformed `find_library` when `VCPKG_LIBRARY_LINKAGE != VCPKG_CRT_LINKAGE`. Do we need a `find_library` overload which transforms the result?

openimageio at least needs `dl`, too

dg0yt commented:
> > So it is more of a general problem than a lapack specific one.
> 
> AFAIU it is a very general problem with the official non-windows triplets in vcpkg, due to the combination of static library linkage with dynamic CRT linkage.

It is good to know this trap. But it is a particular option of the openimageio build system which leads into this trap. (I saw this pattern before in another port: Prepending static suffixes to `CMAKE_FIND_LIBRARY_SUFFIXES`.)

Still more linking issues. And I don't think they are really regressions from my changes: They have been there before but are not visible unless linking executables or enabling certain features - as now done in the vcpkg-ci-openimageio port.

dg0yt committed:
Fix webp usage

dg0yt committed:
Fix static ffmpeg linking

dg0yt committed:
Move config to share/openimageio

dg0yt committed:
More dependency control

dg0yt committed:
Revise ffmpeg patching

dg0yt committed:
Update to 2.1.2

dg0yt committed:
Fix exported pc file

dg0yt committed:
Update versions

dg0yt committed:
Silence warning

dg0yt committed:
Update versions

dg0yt committed:
Catch Imath version conflicts.

This is not a regression but the result of previous opencolorio changes.
It will eventually be resolved by upgrading openexr.

dg0yt committed:
Remove llvm from opencv CI test port [skip actions]

dg0yt committed:
[libheif] Fix symbol export

dg0yt committed:
[skip actions]

JackBoosY commented:
Looks good now?

dg0yt committed:
Update versions

dg0yt commented:
This is good now but based on the opencolorio branch which is offered as a separate PR in #25805.

JackBoosY commented:
So should we merge #25805 first?

dg0yt commented:
> So should we merge #25805 first?

Order doesn't matter, except that #25805 becomes empty when you merge this one first. No merge conflict expected.

dg0yt committed:
Merge remote-tracking branch 'origin/master'

JackBoosY commented on ports/libheif/portfile.cmake:
```diff
@@ -26,10 +26,11 @@ endif()
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/libheif/heif.h" "defined(_MSC_VER) && !defined(LIBHEIF_STATIC_BUILD)" "defined(_WIN32)")
 else()
-    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/libheif/heif.h" "!defined(LIBHEIF_STATIC_BUILD)" "0")
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/libheif/heif.h" "defined(_MSC_VER) && !defined(LIBHEIF_STATIC_BUILD)" "0")
```
Should this be the following to match line 29?
```
    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/libheif/heif.h" "defined(_MSC_VER) && !defined(LIBHEIF_STATIC_BUILD)" "!defined(_WIN32)")
```

dg0yt replied:
No. Line 29 is `defined(_WIN32) && 1`. This line is `defined(_WIN32) && 0`. 
(The pattern controls `declspec(export)`, and _MSC_VER must be replaced for mingw.)

JackBoosY commented on ports/openimageio/fix-static-ffmpeg.patch:
```diff
@@ -0,0 +1,18 @@
++    include(CheckLinkerFlag)
++    check_linker_flag(C "LINKER:-Bsymbolic" supports_bsymbolic)
++    if(supports_bsymbolic)
++        add_link_options("LINKER:-Bsymbolic")
```
This compiler flag should be passed by ffmpeg, right?

dg0yt replied:
Ideally, it would be in the exported usage requirements of ffmpeg.

JackBoosY commented:
@Neumann-A Do you have any questions about this change?

Neumann-A commented on ports/openimageio/portfile.cmake:
```diff
@@ -1,25 +1,27 @@
-file(MAKE_DIRECTORY "${SOURCE_PATH}/ext/robin-map/tsl")
-
-string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" LINKSTATIC)
+file(REMOVE
+    "${SOURCE_PATH}/src/cmake/modules/FindFFmpeg.cmake"
+    "${SOURCE_PATH}/src/cmake/modules/FindLibheif.cmake"
+    "${SOURCE_PATH}/src/cmake/modules/FindLibRaw.cmake"
+    "${SOURCE_PATH}/src/cmake/modules/FindLibsquish.cmake"
+    "${SOURCE_PATH}/src/cmake/modules/FindOpenCV.cmake"
+    "${SOURCE_PATH}/src/cmake/modules/FindOpenJPEG.cmake"
+    "${SOURCE_PATH}/src/cmake/modules/FindWebP.cmake"
+)
```
It probably would have been port update friendly if those wouldn't be all removed but patched to do what you did to make them work correctly. Installing port dependent find modules is in general not a problem if all return a similar enough result.

dg0yt replied:
In the most ports and updates I observed, the changes are made where the package is used, not in the vendored Find module.
And I'm not convinced that patching find modules is more update friendly than patching.
Also, I have already seen problems with custom find modules interfering with transitive dependencies.

Neumann-A commented:
> @Neumann-A Do you have any questions about this change?

Is the `m` linkage solved or do we just assume it works somehow?

dg0yt commented:
> Is the `m` linkage solved or do we just assume it works somehow?

It is resolved by not using `LINKSTATIC=ON`. It would modify the library search suffix list.

(More information: This option does also toggle definitions designed to mark DLL imports from dependencies. In vcpkg, we don't need it when we fix it in the dependencies' headers (-> libheif update). Alternative: Keep passing `LINKSTATIC=ON`, but only patch out the suffix list code.)

dg0yt commented:
Any objections to move to reviewed?

Adela0814 commented:
I encountered the following issue when testing feature `opencolorio`:

Repro step: run `.\vcpkg.exe install openimageio[opencolorio]:x64-windows`

Error message in `\vcpkg\buildtrees\openimageio\config-x64-windows-out.log`:
```
CMake Error at src/cmake/externalpackages.cmake:172 (message):
  OpenColorIO and OpenEXR use incompatible versions of Imath.  You cannot use
  openimageio[opencolorio] for this configuration.
Call Stack (most recent call first):
  CMakeLists.txt:141 (include)
```

dg0yt commented:
I think I made this error message very clear.  

It is not a regression. It is not fixable until openexr3 is available in vcpkg. But user can (try to) select compatible versions in manifest mode.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
Add baseline =pass entries.

BillyONeal reviewed (approved):
I really don't claim to understand everything you've done here, but this port seems like it has .... a lot of issues before you showed up, and you're adding much needed testing, thank you!

Added some =pass baseline entries, will merge once build comes back. Thanks!

dg0yt commented:
This port is just illustrating problems with versions, features and testing in vcpkg. When opencolorio and imath were updated, no testing was done for openimageio[opencolorio]. And there is also the link to openexr, another versioning and dependency disaster, with two starved PR attempts...

BillyONeal commented:
:sigh: Our friend the intermittent x264 build failure on osx is here :/

BillyONeal commented:
/azp run

ras0219-msft commented:
Thanks for the PR and all the fixes (as always)!

We would like to improve testing of features, such as perhaps at least testing each feature of each changed port in the PR build.

dg0yt commented:
> We would like to improve testing of features, such as perhaps at least testing each feature of each changed port in the PR build.

Testing each feature isolated and in combination? That's a lot a to build. And these blind spots remain:
- running the package's native tests (ruled out by policy)
- systematically testing exported config (but somewhat covered through downstream ports, including `vpckg-ci-xxx`)
- systematically testing downstream ports with opt-in dependency (that's how port opencolorio became incompatible with openimageio already before this PR).
